Add ability get the base structure of a valid metadata item from `datalad-catalog`

jsheunis opened the pull request:
- This removes the now-redundant code from this repo, and deduplicates code across repositories
- Equivalent of https://github.com/sfb1451/tabby-utils/pull/9
- Closes https://github.com/psychoinformatics-de/sfb1451-projects-catalog/issues/52

jsheunis committed:
Add ability get the base structure of a valid metadata item from datalad-catalog

This removes the now-redundant code from this repo,deduplicates code across repositories

mslw commented on code/inject_metadata.py:
```diff
@@ -13,49 +13,26 @@
+        dataset_version=ds.repo.get_hexsha(),
+        source_name="manual_addition",
+        source_version="0.1.0",
+    )
```
This effectively removes (dataset) `name` property from the generated `meta_item`:
-  datalad/datalad-catalog/pull/378 that added `get_metadata_item` also removed `name` from the list of required fields
- overall, I think that's a good idea
- however, AFAIK, `catalog-add` verifies the metadata item againist the copy of the schema stored by the catalog instance, not the one in datalad-catalog
- therefore the instance's schema needs to change as part of the PR, otherwise there will be errors

jsheunis replied:
very good point!

mslw commented on code/list_files.py:
```diff
@@ -31,10 +34,15 @@ def is_file(res):
 
 ds = Dataset(args.dataset)
-file_required_meta = {
-    "dataset_id": ds.id,
-    "dataset_version": ds.repo.get_hexsha(),
-}
+
+file_required_meta = get_metadata_item(
+    item_type='file',
+    dataset_id=ds.id,
+    dataset_version=ds.repo.get_hexsha(),
+    source_name="manual_addition",
+    source_version="0.1.0",
+    exclude_keys=["path"],
+)
```
Compared to the previous minimalist approach, this would add `metadata sources` in its non-mandatory entirety (`agent_name`, `agent_email`, `source_time`) -- not sure if intentional, but `get_metadata_item` [always generates metadata sources with not-just-the-required properties](https://github.com/datalad/datalad-catalog/blob/ba656ce3897b46524d2f82f9a30bd999d1a14a73/datalad_catalog/schema_utils.py#L132-L135).

I agree that this would produce a more complete metadata record.

But repeat that over 10 000 files, and it's about half a megabyte used only for spelling out my name & email over and over again. Maybe It shouldn't bother me (see paragraph above) but it slightly does.

jsheunis replied:
Personally, I don't mind if your name appears 10k times :) but I agree this is worth reconsidering

mslw reviewed (commented):
Thanks for the PR!

I didn't have a chance to review the changes elsewhere, so I'll post a general comment here.

This might not be a very big gain in terms of lines of code saved, but I agree that it feels nicer to look up a function signature (instead of schema definition), and it is more convenient to not have to remember the structure of `metadata_sources` object.

However, I feel that this change is tricky for two reasons that I tried to explain in the code comments.

mslw committed:
Update schema to match catalog development

This changes the required fields for dataset (-name +metadata_sources)
and file (+metadata_sources), matching changes introduced in (1)

In addition to aligning the schema with current development state of
datalad-catalog, this should allow validating metadata generated based
on catalog schema (e.g. using schema_utils, which was added together
with the described schema change) againist the catalog instance's
schema.

(1) github.com/datalad/datalad-catalog/commit/b8e38ad22e76c08729d3393f06655c1b65344c3f

mslw commented:
I pushed a schema update to this PR, hope you don't mind.

I think we'll go for this change.
